aries/aries_vcx_core/src/wallet2/mod.rs

```diff
@@ -0,0 +1,393 @@
+use async_trait::async_trait;
+use derive_builder::Builder;
+#[cfg(feature = "vdrtools_wallet")]
+use indy_api_types::domain::wallet::Record as IndyRecord;
+use serde::{Deserialize, Serialize};
+
+use self::entry_tag::EntryTags;
+use crate::{errors::error::VcxCoreResult, wallet::structs_io::UnpackMessageOutput};
+
+#[cfg(feature = "vdrtools_wallet")]
+pub mod indy_wallet;
+
+pub mod entry_tag;
+pub mod utils;
+
+pub struct Key {
+    pub pubkey_bs58: String,
+}
+
+#[derive(Debug, Default, Clone, Builder)]
+pub struct Record {
+    pub category: String,
+    pub name: String,
+    pub value: String,
+    #[builder(default = "EntryTags::default()")]
+    pub tags: EntryTags,
+}
+
+#[cfg(feature = "vdrtools_wallet")]
+impl From<IndyRecord> for Record {
+    fn from(ir: IndyRecord) -> Self {
+        Self {
+            name: ir.id,
+            category: ir.type_,
+            value: ir.value,
+            tags: ir.tags.into(),
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct DidData {
+    did: String,
+    verkey: String,
+}
+
+pub enum SearchFilter {
+    JsonFilter(String),
+}
+
+#[async_trait]
+pub trait BaseWallet2: RecordWallet + DidWallet {}
+
+#[async_trait]
+pub trait DidWallet {
+    async fn create_and_store_my_did(
+        &self,
+        seed: Option<&str>,
+        method_name: Option<&str>,
+    ) -> VcxCoreResult<DidData>;
+
+    async fn did_key(&self, name: &str) -> VcxCoreResult<String>;
+
+    async fn replace_did_key_start(&self, did: &str, seed: Option<&str>) -> VcxCoreResult<String>;
+
+    async fn replace_did_key_apply(&self, did: &str) -> VcxCoreResult<()>;
+
+    async fn sign(&self, key: &str, msg: &[u8]) -> VcxCoreResult<Vec<u8>>;
+
+    async fn verify(&self, key: &str, msg: &[u8], signature: &[u8]) -> VcxCoreResult<bool>;
+
+    async fn pack_message(
+        &self,
+        sender_vk: Option<String>,
+        receiver_keys: Vec<Key>,
+        msg: &[u8],
+    ) -> VcxCoreResult<Vec<u8>>;
+
+    async fn unpack_message(&self, msg: &[u8]) -> VcxCoreResult<UnpackMessageOutput>;
+}
+
+#[async_trait]
+pub trait RecordWallet {
+    async fn add_record(&self, record: Record) -> VcxCoreResult<()>;
+
+    async fn get_record(&self, name: &str, category: &str) -> VcxCoreResult<Record>;
+
+    async fn update_record_tags(
+        &self,
+        name: &str,
+        category: &str,
+        new_tags: EntryTags,
```

Comment:
Can somebody please explain the concepts of name, category and tags, please? Are we using them anywhere? I assume they are specific for indy, libvdrtools and askar wallets (not for every conceivable wallet impl)?

Reply:
Afaik `category + name` is askar vocabulary which maps to vdrtools concept of "type + id". Semantically the same thing, pair together representing a key in the wallet. In vdrtools terms there's example in `credx_anoncreds.rs` 
```
        wallet
            .add_wallet_record(CATEGORY_CRED_DEF, &cred_def_id.0, &str_cred_def, None)
            .await?;
```
- `category`/ `type` value is `VCX_CRED_DEF`
- `name` / `id` is some random Id
- `tags` seems to be key-value pair in `vdrtools`, never used it myself. @xprazak2 mentioned that concept of tags exists in askar too, I presume without much changes.  Given it's a key-value (and the value can be optionally encrypted, in case of sensitive data) I think one could for example store `date: 05.01.2024` key-value pair. Unless I am wrong, "tags" might be a bit of a misnomer from old times, and perhaps "attributes" would better reflect the nature of the data.